assume/common/forecasts.py

```diff
@@ -359,39 +365,61 @@ def calculate_market_price_forecast(self, market_id):
 
+        # 2. Calculate marginal costs for each unit and time step.
+        #    The resulting DataFrame has rows = time steps and columns = units.
         marginal_costs = powerplants_units.apply(self.calculate_marginal_cost, axis=1).T
-        sorted_columns = marginal_costs.loc[self.index[0]].sort_values().index
+
+        # 3. Get forecast availabilities and reformat the column names to match unit identifiers.
         col_availabilities = self.forecasts.columns[
             self.forecasts.columns.str.startswith("availability")
         ]
-        availabilities = self.forecasts[col_availabilities]
+        availabilities = self.forecasts[col_availabilities].copy()
         availabilities.columns = col_availabilities.str.replace("availability_", "")
 
+        # 4. Compute available power for each unit at each time step.
+        #    Since max_power is a float, this multiplication broadcasts over each column.
         power = self.powerplants_units.max_power * availabilities
-        cumsum_power = power[sorted_columns].cumsum(axis=1)
 
+        # 5. Process the demand.
+        #    Filter demand units with a bidding strategy and sum their forecasts for each time step.
         demand_units = self.demand_units[
             self.demand_units[f"bidding_{market_id}"].notnull()
         ]
         sum_demand = self.forecasts[demand_units.index].sum(axis=1)
 
-        # initialize empty price_forecast
-        price_forecast = pd.Series(index=self.index, data=0.0)
-
-        # start with most expensive type (highest cumulative power)
-        for col in sorted_columns[::-1]:
-            # find times which can still be provided with this technology
-            # and cheaper once
-            cheaper = cumsum_power[col] > sum_demand
-            # set the price of this technology as the forecast price
-            # for these times
-            price_forecast.loc[cheaper] = marginal_costs[col].loc[cheaper]
-            # repeat with the next cheaper technology
+        # 6. Initialize the price forecast series.
+        price_forecast = pd.Series(index=self.index, data=np.nan)
+
+        # 7. Loop over each time step
+        for t in self.index:
+            # Get marginal costs and available power for time t (both are Series indexed by unit)
+            mc_t = marginal_costs.loc[t]
+            power_t = power.loc[t]
+            demand_t = sum_demand.loc[t]
+
+            # Sort units by their marginal cost in ascending order for time t.
+            sorted_units = mc_t.sort_values().index
+            sorted_mc = mc_t.loc[sorted_units]
+            sorted_power = power_t.loc[sorted_units]
+
+            # Compute the cumulative sum of available power in the sorted order.
+            cumsum_power = sorted_power.cumsum()
+
+            # Find the first unit where the cumulative available power meets or exceeds demand.
+            matching_units = cumsum_power[cumsum_power >= demand_t]
+            if not matching_units.empty:
```

Comment:
```suggestion
            if matching_units.empty:
```

I would turn this around to avoid the double negation in the else case.
And but the price=1000 case at first